packages/react-devtools-shared/src/utils.js

```diff
@@ -305,3 +323,277 @@ export function setInObject(
+      return 'Fragment';
+    case Lazy:
+      return 'Lazy';
+    case Memo:
+      return 'Memo';
+    case Portal:
+      return 'Portal';
+    case Profiler:
+      return 'Profiler';
+    case StrictMode:
+      return 'StrictMode';
+    case Suspense:
+      return 'Suspense';
+    default:
+      const {type} = element;
+      if (typeof type === 'string') {
+        return type;
+      } else if (type != null) {
+        return getDisplayName(type, 'Anonymous');
+      } else {
+        return 'Element';
+      }
+  }
+}
+
+const STRINGIFY_OPTIONS = {
+  inlineCharacterLimit: 50,
+  singleQuotes: false,
+};
+
+function truncateForDisplay(
+  string: string,
+  length: number = STRINGIFY_OPTIONS.inlineCharacterLimit,
+) {
+  if (string.length > length) {
+    return string.substr(0, length) + '…';
+  } else {
+    return string;
+  }
+}
+
+// Attempts to mimic Chrome's inline preview for values.
+// For example, the following value...
+//   {
+//      foo: 123,
+//      bar: "abc",
+//      baz: [true, false],
+//      qux: { ab: 1, cd: 2 }
+//   };
+//
+// Would show a preview of...
+//   {foo: 123, bar: "abc", baz: Array(2), qux: {…}}
+//
+// And the following value...
+//   [
+//     123,
+//     "abc",
+//     [true, false],
+//     { foo: 123, bar: "abc" }
+//   ];
+//
+// Would show a preview of...
+//   [123, "abc", Array(2), {…}]
+export function formatDataForPreview(
+  data: any,
+  showFormattedValue: boolean,
+): string {
+  if (data != null && data.hasOwnProperty(meta.type)) {
+    return showFormattedValue
+      ? data[meta.preview_long]
+      : data[meta.preview_short];
+  }
+
+  const type = getDataType(data);
+
+  switch (type) {
+    case 'html_element':
+      return `<${truncateForDisplay(data.tagName.toLowerCase())} />`;
+    case 'function':
+      return truncateForDisplay(data.name);
+    case 'string':
+      return `"${data}"`;
+    case 'bigint':
+      return truncateForDisplay(data.toString() + 'n');
+    case 'symbol':
+      return truncateForDisplay(data.toString());
+    case 'react_element':
+      return `<${truncateForDisplay(
+        getDisplayNameForReactElement(data) || 'Unknown',
+      )} />`;
+    case 'array_buffer':
+      return `ArrayBuffer(${data.byteLength})`;
+    case 'data_view':
+      return `DataView(${data.buffer.byteLength})`;
+    case 'array':
+      if (showFormattedValue) {
+        const formatted = [];
+        for (let i = 0; i < data.length; i++) {
+          formatted[i] = formatDataForPreview(data[i], false);
+        }
+        return `[${truncateForDisplay(formatted.join(', '))}]`;
+      } else {
+        const length = data.hasOwnProperty(meta.size)
+          ? data[meta.size]
+          : data.length;
+        return `Array(${length})`;
+      }
+    case 'typed_array':
+      const shortName = `${data.constructor.name}(${data.length})`;
+      if (showFormattedValue) {
+        return `${shortName} [${truncateForDisplay(
+          data
+            .toString()
```

Comment:
Same question. If the data is very large, are we doing unnecessary work by toString’ing it all, only to take a slice later? Should we first take a slice with a reasonable limit?